lib/Dialect/FIRRTL/Transforms/LowerDomains.cpp

```diff
@@ -0,0 +1,516 @@
+  /// associations.
+  ObjectOp op;
+
+  /// The index of the optional input port that will be hooked up to a field of
+  /// the ObjectOp.  This port is an instance of the domain-lowered class.  If
+  /// this is created due to an output domain port, then this is nullopt.
+  std::optional<unsigned> inputPort;
+
+  /// The index of the output port that the ObjectOp will be connected to.  This
+  /// port communicates back to the user information about the associations.
+  unsigned outputPort;
+
+  /// A vector of minimal association info that will be hooked up to the
+  /// associations of this ObjectOp.
+  SmallVector<AssociationInfo> associations{};
+};
+
+/// Struct of the two classes created from a domain, an input class (which is
+/// one-to-one with the domain) and an output class (which tracks the input
+/// class and any associations).
+struct Classes {
+  /// The domain-lowered class.
+  ClassOp input;
+
+  /// A class tracking an instance of the input class and a list of
+  /// associations.
+  ClassOp output;
+};
+
+/// Thread safe, lazy pool of constant attributes
+class Constants {
+
+public:
+  Constants(MLIRContext *context) : context(context) {}
+
+  // Return an empty ArrayAttr.
+  ArrayAttr getEmptyArrayAttr() {
+    if (!emptyArrayAttr) {
+      llvm::sys::SmartScopedLock<true> lock(mutex);
+      emptyArrayAttr = ArrayAttr::get(context, {});
+    }
+    return emptyArrayAttr;
+  }
+
+private:
+  /// Construct all the field info attributes.
+  void constructFieldNameAttrs() {
+    if (domainInfoIn)
+      return;
+    llvm::sys::SmartScopedLock<true> lock(mutex);
+    domainInfoIn = StringAttr::get(context, "domainInfo_in");
+    domainInfoOut = StringAttr::get(context, "domainInfo_out");
+    associationsIn = StringAttr::get(context, "associations_in");
+    associationsOut = StringAttr::get(context, "associations_out");
+  }
+
+public:
+  StringAttr getDomainInfoIn() {
+    constructFieldNameAttrs();
+    assert(domainInfoIn);
+    return domainInfoIn;
+  }
+
+  StringAttr getDomainInfoOut() {
+    constructFieldNameAttrs();
+    assert(domainInfoOut);
+    return domainInfoOut;
+  }
+
+  StringAttr getAssociationsIn() {
+    constructFieldNameAttrs();
+    assert(associationsIn);
+    return associationsIn;
+  }
+
+  StringAttr getAssociationsOut() {
+    constructFieldNameAttrs();
+    assert(associationsOut);
+    return associationsOut;
+  }
+
+private:
+  /// Mutex indicating that, when held, allows for mutation.
+  llvm::sys::SmartMutex<true> mutex;
+
+  /// An MLIR context necessary for creating new attributes.
+  MLIRContext *context;
+
+  /// Lazily constructed attributes
+  ArrayAttr emptyArrayAttr;
+  StringAttr domainInfoIn;
+  StringAttr domainInfoOut;
+  StringAttr associationsIn;
+  StringAttr associationsOut;
+};
+
+class LowerModule {
+
+public:
+  LowerModule(FModuleLike &op, const DenseMap<Attribute, Classes> &classes,
+              Constants &constants)
```

Comment:
IF this is eagerly constructed, please make this const.  Lazy is good too, just noting if go other way.